workshops/test/test_security.py

```diff
@@ -0,0 +1,254 @@
+                admin_site = getattr(v.callback, 'admin_site', None)
+
+                # v.callback is always a function, even when the view is
+                # class based. We try to find the class implementing the view.
+                view_class = getattr(v.callback, 'view_class', None)
+                cls = getattr(v.callback, 'cls', None)
+                class_ = view_class or cls or None
+
+                is_function_based_view = (model_admin is None and
+                                          admin_site is None and
+                                          class_ is None)
+
+                if is_function_based_view:
+                    self.assertTrue(access is not None,
+                                    "You have a function based view with"
+                                    "no access control decorator. This view is"
+                                    "probably accessible to every user. If "
+                                    "this is what you want, use "
+                                    "@login_not_required decorator.")
+                    self.assertIn(access, ['only_for_admins',
+                                           'only_for_logged_in',
+                                           'login_not_required'])
+
+                else:  # class based view
+                    self.assertTrue(access is None,
+                                    'It looks like you used access control '
+                                    'decorator on a class based view. Use '
+                                    'mixin instead.')
+
+                    is_model_admin = isinstance(model_admin, ModelAdmin)
+                    is_admin_site = isinstance(admin_site, AdminSite)
+                    is_api_view = class_ is not None and issubclass(class_, APIView)
+                    is_view = class_ is not None and issubclass(class_, View)
+
+                    if is_model_admin or is_admin_site:
+                        pass  # ignore admin views
+                    elif is_api_view:
+                        pass  # ignore REST API views
+                    else:
+                        assert is_view
+
+                        mixins = set(class_.__mro__)
+                        desired_mixins = {OnlyForAdminsMixin,
+                                          LoginNotRequiredMixin}
+                        found = mixins & desired_mixins
+
+                        self.assertNotEqual(
+                            len(found), 0,
+                            "This view lacks access control mixin and is "
+                            "probably accessible to every user. If this is "
+                            "what you want, use LoginNotRequiredMixin.")
+                        self.assertEqual(
+                            len(found), 1,
+                            "You have more than one access control mixin "
+                            "defined in this view.")
```

Comment:
You have a test for no permission mixin, and you have a test for two excluding permission mixins. Why f-b-views don't have the latter test? Because you *can* chain decorators if you want, so…